README.md
# Plant-Tracker
This is a light-weight iPhone app designed to work with a Raspberry Pi 3b+. After hosting a MQTT server, the app and the Pi can be connected with other components to alert the user to water plants. The app and server follow https://www.raspberrypi.org/forums/viewtopic.php?t=196010 and https://anoop4real.medium.com/how-to-send-data-from-ios-to-raspberry-pi-and-an-lcd-display-using-mqtt-764fee6e8fc5 but expand on the general concepts presented in the tutorials.


# Dependencies
iOS 14.3 or higher | CocoaPods - https://guides.cocoapods.org/using/getting-started.html | CocoaMQTT | Python 3 or higher - https://www.python.org/downloads/ | pip (used to install packages and APIs) - https://pip.pypa.io/en/stable/installing/ | Paho-MQTT - https://pypi.org/project/paho-mqtt/


# MQTT Server Setup
See guide linked above for more in-depth instruction.

Installing MQTT--

```
wget http://repo.mosquitto.org/debian/mosquitto-repo.gpg.key

sudo apt-key add mosquitto-repo.gpg.key

cd /etc/apt/sources.list.d/

sudo wget http://repo.mosquitto.org/debian/mosquitto-buster.list

sudo apt-get update

sudo apt-get install mosquitto
```


Starting the MQTT Server--

```
mosquitto -v
```

When trying to run the server, the error message "Address already in use" might come up. To get around this:

```
sudo lsof -i TCP:1883

sudo kill <the PID of the process running on port 1883>

mosquitto -v
```


# RPI Python Files
On the Raspberry Pi, install the "subscriber.py" file inside the "Python Files" folder. This file will read and decode messages received from the client device.

Make sure the required libraries are already installed

```
sudo apt-get install python-dev

sudo apt-get install python-rpi.gpio

sudo apt-get install python-pip

pip install paho-mqtt
```

Compile the python file as an executable, then run it

```
cd <location of host.py file>

chmod +x host.py

./host.py
```



# Sensor Setup (Software)
The recommended sensors for this project are Adafruit STEMMA Soil Sensors (https://learn.adafruit.com/adafruit-stemma-soil-sensor-i2c-capacitive-moisture-sensor). On the RPI, a few packages will need to be installed. The full list of directions for setup can be found here: https://learn.adafruit.com/circuitpython-on-raspberrypi-linux/installing-circuitpython-on-raspberry-pi and here: https://learn.adafruit.com/adafruit-stemma-soil-sensor-i2c-capacitive-moisture-sensor/python-circuitpython-test.

Installing blinka

```
// First run the following command in a command-line
sudo pip3 install --upgrade setuptools


// Next, put the 4 following lines into a file and compile it as a binary. Run this file using "./<file name>"
cd ~

sudo pip3 install --upgrade adafruit-python-shell

wget https://raw.githubusercontent.com/adafruit/Raspberry-Pi-Installer-Scripts/master/raspi-blinka.py

sudo python3 raspi-blinka.py
```

While installing, you may need to update to Python 3. If this is the case, blinka will prompt you to continue (enter "y" to confirm and continue with installation). Python and blinka may take some time to install and update -- once done, you will be prompted to reboot, confirm by entering "y".

After installing blinka, run the below line to install the CircuitPython library

```
sudo pip3 install adafruit-circuitpython-seesaw
```



# Sensor Setup (Hardware)
This project is currently all wired (although solutions for wireless communication between the RPI and sensors are planned). The project uses a breadboard and a T-connecter to allow multiple sensors to be connected to the GPIO pins on the RPI.

In order to connect more than one sensor, the i2c addresses of some sensors must be changed (this is done through a change in the hardware of the sensors that is explained further on Adafruit's website). Once completed, multiple sensors (up to 4) can be used together.


CHANGELOG.md
# Plant-Tracker Changelog

Note: The most recent versions of the app will also be at the top of ViewController.swift as comments

Project created 1/18/2021 -- Changelog begin:

PRE-RELEASES--

	version		 date						changes
	-------		--------		----------------------------------
	pre-0.0.0	2/8/21			Create Xcode files
	
	pre-1.0.0	2/10/21			First working version of Plant Tracker
	
	pre-2.0.0	2/11/21			Changes in this version:
									-Added app icons
									-Added printMessage function and button to request plant data from RPI
									-Changed layout of elements in the storyboard
									
	pre-2.1.0	2/12/21			Changes in this version:
									-Added displayClearRect and displayText functions
									-Added a method of pinging the server to ensure a connection was made
									
	pre-2.2.0	2/12/21			Changes in this version:
									-Changed the way the client pings the server to establish a connection
									
	pre-2.3.0	2/13/21			Changes in this version:
									-Added alert system to addPlant function
									-Tweaked the way the server connection status is displayed
									
	pre-2.4.0	2/13/21			Changes in this version:
									-Added decodeIncomingMsg and publishOutgoingMsg functions
									-Changed the way messages are sent and received
									
	pre-2.5.0	2/13/21			Changes in this version:
									-UI elements will now correctly reposition and resize depending on the device being used
									
	pre-2.5.1	2/13/21			Changes in this version:
									-Documentation cleanup
									-Added line in app between server information and plant information
									
	pre-2.6.0	2/13/21			Changes in this version:
									-Added in support for changing the host IP address
									
	pre-2.6.1	2/13/21			Changes in this version:
									-Fixed UILabel text alignment
									
	pre-3.0.0	2/14/21			Changes in this version:
									-Changed the way data is handled and stored on the server-side
									-Changed the outgoing and incoming message functions on both the iOS and server-side
									-Added in DOCUMENTATION.md to provide clear documentation and conventions
									-Added the "request" or "respond" argument to all messages
									
	pre-3.0.1	2/14/21			Changes in this version:
									-Fixed the way data is handled on server-side
									-Updated documentation; added TO-DO list
									
	pre-3.1.0	2/14/21			Changes in this version:
									-Added support for hashes on the iOS side
									
	pre-3.2.0	2/15/21			Changes in this version:
									-Added error handling on the server and iOS side
									-Updated documentation
									
	pre-3.3.0	2/15/21			Changes in this version:
									-Fixed issues with the server-side data structure
									-Changed the way responses are handled within the app
									
	pre-3.4.0	2/15/21			Changes in this version:
									-Changed host name in a comment (for M.U.)
									-Fixed error checking
									
	pre-3.4.1	2/16/21			Changes in this version:
									-Changed name of python file subscriber.py -> host.py
									-Documentation changes
									-Began implementation of add plant button
									
	pre-3.5.0	2/17/21			Changes in this version:
									-Plant information will now be saved
									-Added userdata folder with user .json files
									
	pre-4.0.0	2/17/21			Changes in this version:
									-Improved and refined plant info data-structure
									-Implemented RES_plantInfoOnStartup
									-Fixed existing issues with plant info not being returned properly
									-Improved error checking; error alerts will now appear
									-The user now has a limited number of plants, and the plant boxes appear properly
									-Plant names appear in the boxes
									-Updated documentation
									
	pre-4.0.1	2/18/21			Changes in this version:
									-Changed the style of the plant boxes
									-Added in the red-green-red gradient bars (not yet functional)
									
	pre-4.0.2	2/18/21			Changes in this version:
									-Minor improvements in error handling
									
	pre-4.1.0	2/18/21			Changes in this version:
									-Plant interaction has been added
									-Plants can now be deleted
									
	pre-4.1.1	2/18/21			Changes in this version:
									-Removed debug statements
									-Spelling correction
									
	pre-4.2.0	2/20/21			Changes in this version:
									-Clicking on a plant will now display sensor information
									-Began introducing the ability to edit plant name
									-Documentation changes
									
	pre-4.2.1	2/20/21			Changes in this version:
									-Code cleanup; added refreshPlantsDisplayed function
									
	pre-5.0.0	2/28/21			Changes in this version:
									-Major data structure changes
									-sensors.py added to read sensor data
									-Sensor functionality added
									-Sensor data will display when clicking on a plant

	pre-5.0.1	2/28/21			Changes in this version:
									-Fixed a bug with sensor data being duplicated
									
	pre-5.1.0	3/1/21			Changes in this version:
									-Fixed a bug with multiple plants of the same name; user's can no longer have multiple plants with the same name
									-Average moisture between all a plant's sensors now displays above the moisture bar
									
	pre-5.1.1	3/6/21			Changes in this version:
									-Small bug fixes
									-Added character limit to plant names
									
	pre-5.1.2	3/6/21			Changes in this version:
									-Moved some elements of the UI
									-Added current version (current not used, but planned to be checked by the server to make sure a given client is acceptable)
									
	pre-5.1.3	3/9/21			Changes in this version:
									-Small patches following pre-5.1.2
									
	pre-5.1.4	3/10/21			Changes in this version:
									-Changed clientName from UIDevice.current.name to UIDevice.current.identifierForVendor.uuidString
									
	pre-5.2.0	3/14/21			Changes in this version:
									-Moisture values now displayed in a 1-10 scale
									-Moisture "bars" now visually display the value

Python Files/host.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import paho.mqtt.client as mqtt # Import mqtt
import paho.mqtt.publish as publish # Import mqtt
from os import path # Import the path function to check if a userdata file exists or not
import json # Used to read and write userdata
import sensors # Import the sensors.py file
import time # Import time
import threading # Import thread
# import RPi.GPIO as gpio # Import GPIO --> This will not be used in this case


serverAddress = "172.20.8.47" # IP address of the RPI
serverTo = "rpi/torpi" # From client to server stream
serverFrom = "rpi/fromrpi" # From server to client stream
serverName = "Host-RPI3B+" # Emily for short

maxPlants = 7 # Maximum number of plants the user can have
maxPlantName = 15 # Longest plant name the user can have

errCausingHash = "null" # If an error occurs, this is sent back to the client who caused the error for debug

kickHash = {} # Used to keep track of each client and the number of messages they have sent
markEjection = [] # Used to keep track of offending clients
monitorInterval = 15 # How often the server completes a cleanup routine to kick clients sending too many messages
msgThreshold = 15 # If the client sends more than this many messages in monitorInterval time, kick them


# ======================================================================
# Request Tag Functions---


# ======================================================================
# def REQ_plantSensorData
#
def REQ_plantSensorData(msg):
    
    userpath = "userdata/" + msg["sender"] + ".json"
    # Check for existing file
    if (path.exists(userpath)):
        try:
            with open(userpath) as infile:
                plants = json.load(infile)
        # An error occured while reading plant information from the users's .json file
        except:
            # If the requesting client has no plant data, throw an error
            operationError("ERR_noPlantDataToRequest", errCausingHash, msg["sender"])
            return
                
        # Get the data from the sensors
        if (msg["payload"] == "all"): # the user requested all plant data
            
            # For every sensor of every plant, return the data for each plant at a time
            for plant in plants:
                # Data for any given plant
                dataForPlant = []
                
                # Parse the json sensor names as a literal array
                sensorsParsed = json.loads(plant["Sensors"])
                
                for sensor in range(len(sensorsParsed)):
                    try:
                        sensorData = sensors.readSensor(int(sensorsParsed[sensor], 16))
                    except:
                        operationError("ERR_invalidPlantSensorID", errCausingHash, msg["sender"])
                        return
                    
                    # If the sensor ID is invalid, throw an error
                    if (sensorData == "ERR_invalidPlantSensorID"):
                        operationError("ERR_invalidPlantSensorID", errCausingHash, msg["sender"])
                        return
                        
                    # If the data was valid, save it
                    sensorData["plant"] = plant["Name"]
                    sensorData["sensor"] = sensorsParsed[sensor]
                    sensorData = str(sensorData).replace(":", "--")
                    dataForPlant = sensorData
                    
                    # Otherwise, return the data
                    publishOutgoingResponse("0", serverName, msg["sender"], str(dataForPlant), "RES_plantSensorData")
# end: def REQ_plantSensorData

# ======================================================================
# def REQ_numPlants
#
def REQ_numPlants(msg):
    # Make sure the user has plant data already
    userpath = "userdata/" + msg["sender"] + ".json"
    
    # Check for existing file
    if (path.exists(userpath)):
        try:
            with open(userpath) as infile:
                plants = json.load(infile)
            # Send the client back the number of plants
            publishOutgoingResponse("0", serverName, msg["sender"], str(len(plants)), "RES_numPlants")
        except:
            # If the requesting client has no plant data, throw an error
            operationError("ERR_noPlantDataToRequest", "null", msg["sender"])
            return
# end: def REQ_numPlants

# ======================================================================
# def REQ_plantInfoOnStartup
#
def REQ_plantInfoOnStartup(msg):
    # Make sure the user has plant data already
    userpath = "userdata/" + msg["sender"] + ".json"
    
    # Check for existing file
    if (path.exists(userpath)):
        with open(userpath) as infile:
            plants = json.load(infile)
    # If the requesting client has no plant data, throw an error
    else:
        operationError("ERR_noPlantDataToRequest", "", msg["sender"])
        return
        
    # Get the specific plant to return
    plantDataToSend = plants[int(msg["payload"])]
    plantDataAsStr = json.dumps(plantDataToSend)
    plantDataAsStr = plantDataAsStr.replace(":", "--")
    # Return the data for a single plant
    publishOutgoingResponse("0", serverName, msg["sender"], plantDataAsStr, "RES_plantInfoOnStartup")
# end: def REQ_plantInfoOnStartup

# ======================================================================
# def REQ_addNewPlant
#
def REQ_addNewPlant(msg):
    # Init the path of the new or existing file
    userpath = "userdata/" + msg["sender"] + ".json"
    userdata = msg["payload"].split("--")
    
    # Make sure the path exsits
    if (path.exists(userpath)):
        # First read in any existing data
        try:
            with open(userpath) as infile:
                plants = json.load(infile)
        except:
            plants = []
    # If the path doesn't exist create new data
    else:
        plants = []
        
    # The user already has the maximum number of plants
    if (len(plants) >= maxPlants):
        operationError("ERR_tooManyPlants", errCausingHash, msg["sender"])
        return
    # The user already has a plant with this name
    for i in range(len(plants)):
        if (userdata[0] == plants[i]["Name"]):
            operationError("ERR_plantNameTaken", errCausingHash, msg["sender"])
            return
    # Make sure the plant name isn't too long
    if (len(userdata[0]) > maxPlantName):
        operationError("ERR_plantNameTooLong", errCausingHash, msg["sender"])
        return
        
    # Init the data to save
    plants.append({
        "Name"      :   userdata[0],
        "Sensors"   :   userdata[1]
    })
    
    # Save the data (this will create a new file if one does not already exist)
    with open(userpath, "w") as outfile:
        json.dump(plants, outfile)
    
    # Print out what was saved
    print("New plant added with data: " + msg["payload"] + ", for user: " + msg["sender"])
# end: def REQ_addNewPlant

# ======================================================================
# def REQ_deletePlant
#
def REQ_deletePlant(msg):
    # Make sure the user has plant data already
    userpath = "userdata/" + msg["sender"] + ".json"
    
    # Check for existing file
    if (path.exists(userpath)):
        with open(userpath) as infile:
            plants = json.load(infile)
    # If the requesting client has no plant data, throw an error
    else:
        operationError("ERR_noPlantDataToRequest", errCausingHash, msg["sender"])
        return
        
    # Search for and remove the desired plant
    for i in range(len(plants)):
        # If the plant was found, delete it
        if (plants[i]["Name"] == msg["payload"]):
            del plants[i]
            
            # Save the new plant data (any not deleted plants)
            with open(userpath, "w") as outfile:
                json.dump(plants, outfile)
            return
            
        # If the plant was not found, throw an error
        if (i >= len(plants) - 1):
            operationError("ERR_cannotDeletePlant", errCausingHash, msg["sender"])
# end: def REQ_deletePlant

# ======================================================================
# def REQ_editPlant
#
def REQ_editPlant(msg):
    # Make sure the user has plant data already
    userpath = "userdata/" + msg["sender"] + ".json"
    
    # Get the name of the plant to be changed and the new data
    plantToChange = msg["payload"].split(",")
    oldName = plantToChange[0] # Old plant name, the plant to look for
    newName = plantToChange[1] # New name for that plant
    
    # Check for existing file
    if (path.exists(userpath)):
        with open(userpath) as infile:
            plants = json.load(infile)
    # If the requesting client has no plant data, throw an error
    else:
        operationError("ERR_noPlantDataToRequest", errCausingHash, msg["sender"])
        return
    
    # Make sure the user isn't renaming a plant to an already existing name
    for i in range(len(plants)):
        # If the names are equal, the user probably pressed "save" by accident, so don't give them an error
        if (newName == oldName):
            return
        # If a different plant than the one being changed has the newName, throw an error
        if (plants[i]["Name"] == newName):
            operationError("ERR_plantNameTaken", errCausingHash, msg["sender"])
            return
        
    # Search for and change the desired plant
    for i in range(len(plants)):
        # If the plant was found, edit it accodingly
        if (plants[i]["Name"] == oldName):
                
            # If everything is good, change the name
            plants[i]["Name"] = newName
            
            # Save the new plant data (any not deleted plants)
            with open(userpath, "w") as outfile:
                json.dump(plants, outfile)
            return
            
        # If the plant was not found, throw an error
        if (i >= len(plants) - 1):
            operationError("ERR_cannotDeletePlant", errCausingHash, msg["sender"])
# end: def REQ_editPlant



# ======================================================================
# def connectionStatus
#
# Subscribes the RPI to the topic "rpi/to" which handles data from the
# iOS device to the RPI and the topic "rpi/from" which handles data from
# the RPI to the iOS
#
# Arguments--
#
# client:	the client connecting
#
# userdata:	the data from the user connecting
#
# flags:	
#
# rc:
#
# Returns--
#
# None
#
def connectionStatus(client, userdata, flags, rc):
    mqttClient.subscribe(serverTo)
    mqttClient.subscribe(serverFrom)
# end: def connectionStatus


# ======================================================================
# def operationError
#
# Creates and throws an error if something is wrong with a request
#
# Arguments--
#
# error:    the error to throw
#
# msg:  the hash that created the error
#
# receiver: the client who should receiver the error
#
# Returns--
#
# None
def operationError(error, msg, receiver):
    
    publishOutgoingResponse("0", serverName, receiver, msg, error)

# end: def operationError


# ======================================================================
# def publishOutgoingResponse
#
# Publish an outgoing message with correct formatting
#
# Arguments--
#
# msgID:        the ID of the message being sent
#
# clientName:   the name of the client sending the message
#
# payload:      the contents of the message
#
# operation:    the request or response
#
# Returns--
#
# None
#
def publishOutgoingResponse(msgID, sender, receiver, payload, operation):
    newMsg = "ID:" + msgID + ";sender:" + sender + ";receiver:" + receiver + ";payload:" + payload + ";operation:" + operation
    publish.single(serverFrom, newMsg, hostname = serverAddress)
# end: def publishMessage


# ======================================================================
# def monitorClients
#
# Monitors how many messages each client is sending and ejects clients
# sending too many messages
#
# Arguments--
#
# None
#
# Returns--
#
# None
#
def monitorClients():
    while (True):
        for i in globals()['kickHash']:
            if (globals()['kickHash'][i] > msgThreshold):
                # Disconnect the client
                markEjection.append(i)
                
        globals()['kickHash'] = {} # Prepare the kickHash for the next cleanup
        time.sleep(monitorInterval)
# end: def monitorClients

x = threading.Thread(target=monitorClients)
x.start()


# ======================================================================
# def decodeIncomingRequest
#
# Decode a message recieved from the topic
#
# Arguments--
#
# client:	the client connecting
#
# userdata:	the data of the connected user
#
# msg:		the message that came with the topic
#
# Returns--
#
# None
#
def decodeIncomingRequest(client, userdata, msg):
    
    # Whenever a message is sent, complete the cleanup routine by kicking any previously offending clients
    for client in markEjection:
        publishOutgoingResponse("0", serverName, client, "none", "ERR_kickedForSpam")
        globals()['markEjection'] = [] # Reset the eject list
    
    # Decode the message
    entireMsg = msg.payload.decode(encoding='UTF-8')
    globals()['errCausingHash'] = entireMsg.replace(";", "|")
    globals()['errCausingHash'] = globals()['errCausingHash'].replace(":", "--")
    
    # Split the message into its elements
    msgElements = entireMsg.split(";")
    msgHash = {}
    
    # Search through all the elements to make sure they are valid
    for i in msgElements:
        keyValue = i.split(":")

        # Confirm there is one key and one value only
        if (len(keyValue) > 2): 
            operationError("ERR_missingKeys", errCausingHash, msgHash["sender"])
            return
        elif (len(keyValue) < 2): 
            operationError("ERR_missingVals", errCausingHash, msgHash["sender"])
            return

        key = keyValue[0]
        value = keyValue[1]
        
        msgHash[key] = value
    
    # Make sure there is the reqired number of elements in the hash
    if (not (len(msgHash) == 5)): 
        operationError("ERR_hashLength", errCausingHash, msgHash["sender"])
        return
        
    # Check to see if the connecting client is already in kickHash
    if (not msgHash["sender"] == serverName):
        if (msgHash["sender"] in kickHash):
            kickHash[msgHash["sender"]] += 1 # Increment the number of messages they have sent
        else:
            kickHash[msgHash["sender"]] = 1 # If the client did not already exist, add them
        print(f">>MSG CLIENT: " + str(kickHash[msgHash["sender"]]))
    
    # Make sure the message is not just a PUBACK (publish sent back) from the RPI host
    if (not msgHash["sender"] == serverName):
        
        # Hash to handle request tags
        requestTagHash = {
            "REQ_plantSensorData"   :   REQ_plantSensorData,
            "REQ_numPlants"         :   REQ_numPlants,
            "REQ_plantInfoOnStartup":   REQ_plantInfoOnStartup,
            "REQ_addNewPlant"       :   REQ_addNewPlant,
            "REQ_deletePlant"       :   REQ_deletePlant,
            "REQ_editPlant"         :   REQ_editPlant,
        }
        
        # Ignore errors about errors to prevent bouncing back
        dropErr = {
            "ERR_hashLength"		    :	-1,
			"ERR_missingVals"		    :	-2,
			"ERR_missingKeys"		    :	-3,
			"ERR_invalidOpTag"		    :	-4,
            "ERR_noPlantDataToRequest"  :   -5,
            "ERR_tooManyPlants"         :   -6,
            "ERR_cannotDeletePlant"     :   -7,
            "ERR_invalidPlantSensorID"  :   -8,
        }
        
        # Figure out if the request is valid (is it in the hash above?) and call the associated function
        if (msgHash["operation"] in requestTagHash):
            requestTagHash[msgHash["operation"]](msgHash)
        elif (msgHash["operation"] in dropErr):
            print("New ERROR " + msgHash["operation"] + " with payload \"" + msgHash["payload"] + "\". Sender " + msgHash["sender"] + ", Receiver: " + msgHash["receiver"] + ", with ID " + msgHash["ID"])
            return
        else:
            # If the tag is invalid, throw an error
            operationError("ERR_invalidOpTag", errCausingHash, msgHash["sender"])
            return
        
    print("New operation " + msgHash["operation"] + " with payload \"" + msgHash["payload"] + "\". Sender " + msgHash["sender"] + ", Receiver: " + msgHash["receiver"] + ", with ID " + msgHash["ID"])
# end: def decodeIncomingRequest


# Instate Eclipse Paho as mqttClient
mqttClient = mqtt.Client(serverName)

# Set calling functions to mqttClient
mqttClient.on_connect = connectionStatus # Called when the RPI
mqttClient.on_message = decodeIncomingRequest # Called when a message is recieved

# Connect client to Server
mqttClient.connect(serverAddress)
# Monitor client activity forever
mqttClient.loop_forever()
